Root sequence<any> params using CustomAutoRooter

Xanewok opened the pull request:
<!-- Please describe your changes on the following line: -->

Attempt at https://github.com/servo/servo/issues/16678. At the moment this pulls an external [remove-handle-conv](https://github.com/Xanewok/rust-mozjs/tree/remove-handle-conv) branch for rust-mozjs (removing `FromJSValConvertible` implementation for `HandleValue` and adding one for `*mut JSObject`)

In short, this roots the `sequence<any>` and `sequence<object>` function arguments using `CustomAutoRooter` (introduced in https://github.com/servo/rust-mozjs/pull/382).

As per https://github.com/servo/servo/issues/16678#issuecomment-302224110 the underlying vector contains raw gc thing pointers instead of handles. To do that, this introduces new possible `isMember = "AutoRoot"` value in CodegenRust.py.
The rooted argument is passed to a function wrapped in a `CustomAutoRooterGuard` [RAII root guard](https://github.com/servo/rust-mozjs/blob/ed5e37a288b5738d9b571b8100b4a22a2c00f075/src/rust.rs#L586-L588) (e.g. as `CustomAutoRooterGuard<Vec<JSVal>>`)

I felt quite out of my depth when trying to adapt CodegenRust.py code, so I'd appreciate any help with how can be done better/in a more clean manner.
Also the name `CustomAutoRooterGuard` is quite lengthy, so I'm also open to changing it (`AutoRoot`? `AutoRootGuard`?)

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [x] These changes fix #16678  (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19644)
<!-- Reviewable:end -->

Xanewok committed:
Root sequence<{any,object}> IDL arguments using CustomAutoRooter

Also pulls in mozjs 0.1.10 to support the change.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @asajeffrey (or someone else) soon.

jdm reviewed (changes requested):
This is a good start. I'd like to see what it looks like if we don't overload isMember, as well as having a test that verifies that it works correctly.

Xanewok commented:
I addressed the review comments in the new commit. I still have to include a test for that.

@jdm I have a question regarding adding the test.
From what I saw, the automated tests for this should be [here](https://github.com/servo/servo/tree/master/tests/wpt/mozilla/tests/mozilla)? Should I test the gc using `window.gc()` like in [transitionend_safety.html](https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/transitionend_safety.html)?

I figured it'd be good to check if the values inside passed sequence<any> in some test function receiving sequence<any> are not all undefined? Or maybe use specific ones? In the latter case, given how far away the actual tests in html are from testbinding.rs file, I'm not sure the specified value usage will be clear.

By the was, is there a way to run only a specific .html test using `./mach test wpt <test>`? Pointing to the `tests/wpt/mozilla/tests/mozilla` executes the full test suite with 1335 tests, but when I point to a specific .html inside test run starts but doesn't execute a single test.

Xanewok commented:
Added an additional WPT test. Hope this is what you had in mind, didn't touch Servo tests before :thinking:

emilio commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -6421,9 +6436,24 @@ def type_needs_tracing(t):
     info = getJSToNativeConversionInfo(
-        ty, descriptorProvider, isArgument=True)
+        ty, descriptorProvider, isArgument=True,
+        isAutoRooted=True if type_needs_auto_root(ty) else False)
```
nit: just `isAutoRooted=type_needs_auto_root(ty)`?

Xanewok replied:
Right, thanks!

emilio reviewed (commented):
Thanks a lot for finally fixing this! :)

jdm commented on tests/wpt/mozilla/tests/mozilla/custom_auto_rooter.html:
```diff
@@ -0,0 +1,30 @@
+</head>
+<script>
+test(function() {
+  function assert_array_equals(lhs, rhs) {
```
This should already exist: http://web-platform-tests.org/writing-tests/testharness-api.html

Xanewok replied:
Fixed, thanks!

jdm commented on tests/wpt/mozilla/tests/mozilla/custom_auto_rooter.html:
```diff
@@ -0,0 +1,30 @@
+  assert_array_equals(seq, t.anySequencePassthrough(seq));
+
+  var seq = new Array(null, {'inner': 1, 'a': 'as', 'toString': function() { seq = null; return Object.prototype.toString(this); }}, {'inner': 2}, null, undefined);
+  assert_array_equals(seq, t.anySequencePassthrough(seq));
```
Just to verify, have you checked if any of these tests fail without your codegen changes?

Xanewok replied:
Yes. Previously, with https://github.com/Xanewok/servo/commit/01d1039311fd89cec84d6080333e32114fe03f80:
```ts
var seq = [5, "str", { key: "val" }, undefined];
console.log("seq: " + seq);
var lvalue = t.anySequencePassthrough(seq);
console.log("lvalue: " + lvalue);
console.log("rvalue: " + t.anySequencePassthrough(seq));
assert_array_equals(seq, t.anySequencePassthrough(seq), "Returned simple array is the same");

```
errors with:
```
 0:13.20 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27314) "seq: 5,str,[object Object],"   
 0:13.20 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27314) "lvalue: 0,5e-324,5e-324,5e-324"
 0:13.20 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27314) "rvalue: 0,5e-324,5e-324,5e-324"
...
Expected PASS, got FAIL                                 
assert_array_equals: Returned simple array is the same property 0, expected 0 but got 5
```

jdm replied:
\o/

jdm reviewed (changes requested):
The codegen changes look much nicer. Thanks!

jdm commented:
Once https://github.com/servo/rust-mozjs/pull/388 merges then we can update this to not use the branch.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19614) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Upstream PR merged.

Xanewok commented:
r? @jdm

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit a01d1ea has been approved by `jdm`

<!-- @bors-servo r=jdm a01d1eabefeb9ec6a65c518234647ddbf15e7f92 -->

bors-servo commented:
:hourglass: Testing commit a01d1eabefeb9ec6a65c518234647ddbf15e7f92 with merge 989d2fd53267d063212ef3b7b7f13f2402943c4a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10298), [arm32](http://build.servo.org/builders/arm32/builds/10534), [arm64](http://build.servo.org/builders/arm64/builds/10561), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10587), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7439), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7455), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10659), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3073), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3091), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7340), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7400), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3159), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3111), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5959)
Approved by: jdm
Pushing 989d2fd53267d063212ef3b7b7f13f2402943c4a to master...
